plugins/woocommerce/includes/class-wc-discounts.php

```diff
@@ -565,7 +565,9 @@ protected function apply_coupon_remainder( $coupon, $items_to_apply, $amount ) {
 		$total_discount = 0;
 
 		foreach ( $items_to_apply as $item ) {
-			for ( $i = 0; $i < $item->quantity; $i++ ) {
+			$quantity = NumberUtil::ceil( $item->quantity );
```

Comment:
New util takes care of ensuring `ceil` works regardless of what format quantity is in.